keps/sig-auth/0035-20121220-outgoing-webhook-auth.md

```diff
@@ -0,0 +1,122 @@
+approvers:
+  - "@sttts"
+  - "@liggitt"
+editor: TBD
+creation-date: 2018-12-20
+last-updated: 2018-12-20
+status: provisional
+see-also:
+replaces:
+superseded-by:
+---
+
+# Outgoing Webhook Authentication
+
+## Table of Contents
+
+* [Outgoing Webhook Authentication](#outgoing-webhook-authentication)
+  * [Table of Contents](#table-of-contents)
+  * [Summary](#summary)
+  * [Motivation](#motivation)
+      * [Goals](#goals)
+      * [Non-Goals](#non-goals)
+  * [Proposal](#proposal)
+      * [User Stories](#user-stories)
+        * [Story 1](#story-1)
+        * [Story 2](#story-2)
+      * [Implementation Details/Notes/Constraints](#implementation-detailsnotesconstraints)
+      * [Risks and Mitigations](#risks-and-mitigations)
+  * [Graduation Criteria](#graduation-criteria)
+  * [Implementation History](#implementation-history)
+  * [Alternatives](#alternatives)
+
+## Summary
+
+We want to provide a means of authenticating outgoing webhooks from the apiserver and its aggregates.
+
+## Motivation
+
+Outgoing webhooks from the apiserver such as the ones found in [dynamic admission control](https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers) and [dynamic audit](https://kubernetes.io/docs/tasks/debug-application-cluster/audit/#dynamic-backend) 
+suffer from a lack of authentication. The intention of this KEP is to provide a means for the receiving server 
+to authenticate the apiserver and its aggregates.
+
+### Goals
+* A simple means of authenticating apiserver clients.
+
+### Non-Goals
+* Providing all the authentication schemes found in the current kubeconfig.
+
+## Proposal
+
+We propose a simple mechanism for authenticating webhooks using the token [Authentication API](https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/authentication/types.go). The shared 
+[webhook client](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/util/webhook/client.go) will be parameterized to optionally enhance every outgoing request with a token obtained from a  
+[Token Request](https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/authentication/types.go#L112). The receiving server can then check that token using a [Token Review](https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/authentication/types.go#L45).
+
+### User Stories
+
+#### Story 1
+I am a cluster administrator using the dynamic auditing feature and want to make sure the logs I receive are from the apiserver.
+
+#### Story 2
+I am a cluster administrator using dynamic admission control and want to verify that only verified parties are using my token review endpoint.
+
+### Implementation Details/Notes/Constraints
+
+A new struct will be added to the [client config](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/util/webhook/client.go#L40) for outgoing webhooks:
+
+```go
+type AuthInfo struct {
+  TokenRequest  *authenticationv1.TokenRequestSpec
```

Comment:
I'm not sure an entire token request spec makes sense, for a few reasons:

1. the target probably doesn't care about expiration or bound object, just audience

2. it seems safer to make the audience an inherent part of the way the webhook is called (e.g. tokens obtained for calling `https://mywebhook.com` would have an audience of `https://mywebhook.com`, and tokens obtained for calling `myservice` in `mynamespace` would have an audience of `https://myservice.mynamespace.svc`, similar to how [TLS validation for webhook calls is currently done](https://github.com/kubernetes/kubernetes/blob/870c0507272d13b67c4beff34685ff674f716a2d/staging/src/k8s.io/apiserver/pkg/util/webhook/client.go#L151-L158))

Reply:
1. I was thinking the bound object would possibly be used to limit the exposure of a token, but probably isn't necessary

2. Being prescriptive with the audience makes sense. 

So, from the API perspective, we probably just want a boolean switch. I'm thinking it still makes sense to wrap in an an `AuthInfo` struct so we can potentially add other methods down the road.
```go
type AuthInfo struct {
   ProvisionToken  bool
}
type ClientConfig struct {
   Name     string
   URL      string
   CABundle []byte
   AuthInfo *AuthInfo
   Service  *ClientConfigService
}
```